### ISSUE TITLE
Building haddock works on some versions but not others.

### ISSUE THREAD

**Issue by philderbeast** [adamgundry/uom-plugin#67]
Building haddock works on some versions but not others.

``` 
# These work ...
> stack haddock --stack-yaml=stack-8.0.2.yaml
> stack haddock --stack-yaml=stack-8.2.2.yaml
> stack haddock --stack-yaml=stack-8.4.4.yaml

# These don't ...
> stack haddock --stack-yaml=stack-8.6.5.yaml
> stack haddock --stack-yaml=stack-8.8.3.yaml
```

**Comment by philderbeast** [comment]
The 8.6.5 failure.

```
> stack haddock --stack-yaml=stack-8.6.5.yaml
uom-plugin> configure (lib)
Configuring uom-plugin-0.3.0.0...
uom-plugin> build (lib)
Preprocessing library for uom-plugin-0.3.0.0..
Building library for uom-plugin-0.3.0.0..
[13 of 18] Compiling Data.UnitsOfMeasure.Convert [Plugin forced recompilation]
[17 of 18] Compiling Data.UnitsOfMeasure.Defs [Plugin forced recompilation]
uom-plugin> haddock
Preprocessing library for uom-plugin-0.3.0.0..
Running Haddock on library for uom-plugin-0.3.0.0..

src/Data/UnitsOfMeasure/Convert.hs:94:22: error:
    • Could not deduce: (CanonicalBaseUnit b /: CanonicalBaseUnit b)
                        ~ One
        arising from the literal ‘1’
      from the context: HasCanonicalBaseUnit b
        bound by the class declaration for ‘HasCanonicalBaseUnit’
        at src/Data/UnitsOfMeasure/Convert.hs:83:8-27
      or from: Base b ~ CanonicalBaseUnit b
        bound by the type signature for:
                   conversionBase :: forall (proxy :: Symbol -> *).
                                     (Base b ~ CanonicalBaseUnit b) =>
                                     proxy b -> Quantity Rational (Base b /: CanonicalBaseUnit b)
        at src/Data/UnitsOfMeasure/Convert.hs:93:29-122
    • In the expression: 1
      In an equation for ‘conversionBase’: conversionBase _ = 1
    • Relevant bindings include
        conversionBase :: proxy b
                          -> Quantity Rational (Base b /: CanonicalBaseUnit b)
          (bound at src/Data/UnitsOfMeasure/Convert.hs:94:3)
   |
94 |   conversionBase _ = 1
   |                      ^

src/Data/UnitsOfMeasure/Convert.hs:130:22: error:
    • Could not deduce: ((Prod xs /: Prod ys)
                         /: (ListToCBU xs /: ListToCBU ys))
                        ~~ ((Prod xs /: ListToCBU xs) /: (Prod ys /: ListToCBU ys))
        arising from a use of ‘/:’
      from the context: HasCanonical u
        bound by the type signature for:
                   help :: forall (u :: UnitSyntax Symbol).
                           HasCanonical u =>
                           SUnit u -> Quantity Rational (Pack u /: ToCBU u)
        at src/Data/UnitsOfMeasure/Convert.hs:129:1-85
      or from: u ~ (xs ':/ ys)
        bound by a pattern with constructor:
                   SUnit :: forall (xs :: [Symbol]) (ys :: [Symbol]).
                            SList xs -> SList ys -> SUnit (xs ':/ ys),
                 in an equation for ‘help’
        at src/Data/UnitsOfMeasure/Convert.hs:130:7-17
    • In the expression: help' xs /: help' ys
      In an equation for ‘help’:
          help (SUnit xs ys) = help' xs /: help' ys
    • Relevant bindings include
        ys :: SList ys (bound at src/Data/UnitsOfMeasure/Convert.hs:130:16)
        xs :: SList xs (bound at src/Data/UnitsOfMeasure/Convert.hs:130:13)
    |
130 | help (SUnit xs ys) = help' xs /: help' ys
    |                      ^^^^^^^^^^^^^^^^^^^^

src/Data/UnitsOfMeasure/Convert.hs:133:22: error:
    • Could not deduce: (One /: One) ~ One arising from the literal ‘1’
      from the context: AllHasCanonical xs
        bound by the type signature for:
                   help' :: forall (xs :: [Symbol]).
                            AllHasCanonical xs =>
                            SList xs -> Quantity Rational (Prod xs /: ListToCBU xs)
        at src/Data/UnitsOfMeasure/Convert.hs:132:1-98
      or from: xs ~ '[]
        bound by a pattern with constructor: SNil :: SList '[],
                 in an equation for ‘help'’
        at src/Data/UnitsOfMeasure/Convert.hs:133:7-10
    • In the expression: 1
      In an equation for ‘help'’: help' SNil = 1
    |
133 | help' SNil         = 1
    |                      ^

src/Data/UnitsOfMeasure/Convert.hs:134:22: error:
    • Could not deduce: ((Base x *: Prod xs1)
                         /: (CanonicalBaseUnit x *: ListToCBU xs1))
                        ~~ ((Base x /: CanonicalBaseUnit x) *: (Prod xs1 /: ListToCBU xs1))
        arising from a use of ‘*:’
      from the context: AllHasCanonical xs
        bound by the type signature for:
                   help' :: forall (xs :: [Symbol]).
                            AllHasCanonical xs =>
                            SList xs -> Quantity Rational (Prod xs /: ListToCBU xs)
        at src/Data/UnitsOfMeasure/Convert.hs:132:1-98
      or from: (xs ~ (x : xs1), KnownSymbol x)
        bound by a pattern with constructor:
                   SCons :: forall (x :: Symbol) (proxy :: Symbol
                                                           -> *) (xs :: [Symbol]).
                            KnownSymbol x =>
                            proxy x -> SList xs -> SList (x : xs),
                 in an equation for ‘help'’
        at src/Data/UnitsOfMeasure/Convert.hs:134:8-17
    • In the expression: conversionBase p *: help' xs
      In an equation for ‘help'’:
          help' (SCons p xs) = conversionBase p *: help' xs
    • Relevant bindings include
        xs :: SList xs1
          (bound at src/Data/UnitsOfMeasure/Convert.hs:134:16)
        p :: proxy x (bound at src/Data/UnitsOfMeasure/Convert.hs:134:14)
    |
134 | help' (SCons p xs) = conversionBase p *: help' xs
    |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/Data/UnitsOfMeasure/Convert.hs:151:12: error:
    • Could not deduce: v ~~ ((v /: u) *: u) arising from a use of ‘*:’
      from the context: (Fractional a, Convertible u v)
        bound by the type signature for:
                   convert :: forall a (u :: Unit) (v :: Unit).
                              (Fractional a, Convertible u v) =>
                              Quantity a u -> Quantity a v
        at src/Data/UnitsOfMeasure/Convert.hs:150:1-89
    • In the expression:
        (ratio
           (undefined :: proxy' (proxy v)) (undefined :: proxy' (proxy u))
         *:)
      In an equation for ‘convert’:
          convert
            = (ratio
                 (undefined :: proxy' (proxy v)) (undefined :: proxy' (proxy u))
               *:)
    • Relevant bindings include
        convert :: Quantity a u -> Quantity a v
          (bound at src/Data/UnitsOfMeasure/Convert.hs:151:1)
    |
151 | convert = (ratio (undefined :: proxy' (proxy v)) (undefined :: proxy' (proxy u)) *:)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/Data/UnitsOfMeasure/Convert.hs:160:29: error:
    • Could not deduce: (u /: v)
                        ~~ ((u /: ToCBU (Unpack v)) /: (v /: ToCBU (Unpack v)))
        arising from a use of ‘/:’
      from the context: (Fractional a, Convertible u v)
        bound by the type signature for:
                   ratio :: forall a (u :: Unit) (v :: Unit) (proxy :: Unit
                                                                       -> *) (proxy' :: * -> *).
                            (Fractional a, Convertible u v) =>
                            proxy' (proxy u) -> proxy' (proxy v) -> Quantity a (u /: v)
        at src/Data/UnitsOfMeasure/Convert.hs:(157,1)-(159,68)
    • In the second argument of ‘($)’, namely
        ‘conversionRatio (undefined :: proxy u)
           /: conversionRatio (undefined :: proxy v)’
      In the expression:
        fromRational'
          $ conversionRatio (undefined :: proxy u)
              /: conversionRatio (undefined :: proxy v)
      In an equation for ‘ratio’:
          ratio _ _
            = fromRational'
                $ conversionRatio (undefined :: proxy u)
                    /: conversionRatio (undefined :: proxy v)
    • Relevant bindings include
        ratio :: proxy' (proxy u)
                 -> proxy' (proxy v) -> Quantity a (u /: v)
          (bound at src/Data/UnitsOfMeasure/Convert.hs:160:1)
    |
160 | ratio _ _ = fromRational' $ conversionRatio (undefined :: proxy u) /: conversionRatio (undefined :: proxy v)
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Haddock coverage:
Warning: '-' is ambiguous. It is defined
    * in ‘GHC.Num’
    * in ‘GHC.TypeNats’
    You may be able to disambiguate the identifier by qualifying it or
    by hiding some imports.
    Defaulting to '-' defined in ‘GHC.TypeNats’
Warning: '/:' is ambiguous. It is defined
    * at src/Data/UnitsOfMeasure/Internal.hs:182:14
    * at src/Data/UnitsOfMeasure/Internal.hs:88:1
    You may be able to disambiguate the identifier by qualifying it or
    by hiding some imports.
    Defaulting to '/:' defined at src/Data/UnitsOfMeasure/Internal.hs:88:1
Warning: '*:' is ambiguous. It is defined
    * at src/Data/UnitsOfMeasure/Internal.hs:157:14
    * at src/Data/UnitsOfMeasure/Internal.hs:82:1
    You may be able to disambiguate the identifier by qualifying it or
    by hiding some imports.
    Defaulting to '*:' defined at src/Data/UnitsOfMeasure/Internal.hs:82:1
 100% ( 37 / 37) in 'Data.UnitsOfMeasure.Internal'
 100% ( 10 / 10) in 'Data.UnitsOfMeasure.Singleton'
 100% (  3 /  3) in 'Data.UnitsOfMeasure.Show'
 100% (  6 /  6) in 'Data.UnitsOfMeasure.Read'
Warning: 'Unit' is out of scope.
    If you qualify the identifier, haddock can try to link it
    it anyway.
Warning: 'Base' is out of scope.
    If you qualify the identifier, haddock can try to link it
    it anyway.
Warning: 'One' is out of scope.
    If you qualify the identifier, haddock can try to link it
    it anyway.
Warning: 'Unpack' is out of scope.
    If you qualify the identifier, haddock can try to link it
    it anyway.
Warning: 'UnitSyntax' is out of scope.
    If you qualify the identifier, haddock can try to link it
    it anyway.
 100% (  2 /  2) in 'Data.UnitsOfMeasure.Plugin'

--  While building package uom-plugin-0.3.0.0 using:
      /Users/pdejoux/.stack/setup-exe-cache/x86_64-osx/Cabal-simple_mPHDZzAJ_2.4.0.1_ghc-8.6.5 --builddir=.stack-work/dist/x86_64-osx/Cabal-2.4.0.1 haddock --html --hoogle --html-location=../$pkg-$version/ --haddock-option=--hyperlinked-source --haddock-option=--quickjump
    Process exited with code: ExitFailure 1
```

**Comment by philderbeast** [comment]
The 8.8.3 failure that we're asked to report as a GHC bug.

```
> stack haddock --stack-yaml=stack-8.8.3.yaml                                                                          17:24:32
Stack has not been tested with GHC versions above 8.6, and using 8.8.3, this may fail
Stack has not been tested with Cabal versions above 2.4, but version 3.0.1.0 was found, this may fail
uom-plugin-examples> configure
uom-plugin-examples> Configuring uom-plugin-examples-0.1.0.0...
uom-plugin-examples> build
multimap           > configure
uom-plugin-examples> Building uom-plugin-examples-0.1.0.0...
multimap           > Configuring multimap-1.2.1...
multimap           > build
ghc-tcplugins-extra> configure
multimap           > Preprocessing library for multimap-1.2.1..
multimap           > Building library for multimap-1.2.1..
multimap           > [1 of 2] Compiling Data.MultiMap
ghc-tcplugins-extra> Configuring ghc-tcplugins-extra-0.4...
multimap           > [2 of 2] Compiling Data.SetMap
ghc-tcplugins-extra> build
ghc-tcplugins-extra> Preprocessing library for ghc-tcplugins-extra-0.4..
ghc-tcplugins-extra> Building library for ghc-tcplugins-extra-0.4..
ghc-tcplugins-extra> [1 of 1] Compiling GHC.TcPluginM.Extra
multimap           > haddock
multimap           > Preprocessing library for multimap-1.2.1..
multimap           > Running Haddock on library for multimap-1.2.1..
multimap           > Haddock coverage:
multimap           >  100% ( 43 / 43) in 'Data.MultiMap'
multimap           >   96% ( 24 / 25) in 'Data.SetMap'
multimap           >   Missing documentation for:
multimap           >     Module header
multimap           > Documentation created:
multimap           > .stack-work/dist/x86_64-osx/Cabal-3.0.1.0/doc/html/multimap/index.html,
multimap           > .stack-work/dist/x86_64-osx/Cabal-3.0.1.0/doc/html/multimap/multimap.txt
multimap           > copy/register
multimap           > Installing library in /Users/pdejoux/.stack/snapshots/x86_64-osx/861454bd46483e99623797acad7e761a125bed3f708af95b7ac258877dc7020e/8.8.3/lib/x86_64-osx-ghc-8.8.3/multimap-1.2.1-IztW81H8oaY5HG5UUwXVYE
multimap           > Registering library for multimap-1.2.1..
units-parser       > configure
ghc-tcplugins-extra> haddock
ghc-tcplugins-extra> Preprocessing library for ghc-tcplugins-extra-0.4..
ghc-tcplugins-extra> Running Haddock on library for ghc-tcplugins-extra-0.4..
units-parser       > Configuring units-parser-0.1.0.0...
units-parser       > build
units-parser       > Preprocessing library for units-parser-0.1.0.0..
units-parser       > Building library for units-parser-0.1.0.0..
ghc-tcplugins-extra> Haddock coverage:
ghc-tcplugins-extra>  100% ( 18 / 18) in 'GHC.TcPluginM.Extra'
units-parser       > [1 of 1] Compiling Text.Parse.Units
ghc-tcplugins-extra> Documentation created:
ghc-tcplugins-extra> .stack-work/dist/x86_64-osx/Cabal-3.0.1.0/doc/html/ghc-tcplugins-extra/index.html,
ghc-tcplugins-extra> .stack-work/dist/x86_64-osx/Cabal-3.0.1.0/doc/html/ghc-tcplugins-extra/ghc-tcplugins-extra.txt
ghc-tcplugins-extra> copy/register
ghc-tcplugins-extra> Installing library in /Users/pdejoux/.stack/snapshots/x86_64-osx/861454bd46483e99623797acad7e761a125bed3f708af95b7ac258877dc7020e/8.8.3/lib/x86_64-osx-ghc-8.8.3/ghc-tcplugins-extra-0.4-1quW0TFs3k6Ll3d5vMFV56
ghc-tcplugins-extra> Registering library for ghc-tcplugins-extra-0.4..
units-parser       > haddock
units-parser       > Preprocessing library for units-parser-0.1.0.0..
units-parser       > Running Haddock on library for units-parser-0.1.0.0..
units-parser       > Haddock coverage:
units-parser       > Warning: 'Map' is out of scope.
units-parser       >     If you qualify the identifier, haddock can try to link it anyway.
units-parser       >   86% ( 12 / 14) in 'Text.Parse.Units'
units-parser       >   Missing documentation for:
units-parser       >     lex (Text/Parse/Units.hs:165)
units-parser       >     unitStringParser (Text/Parse/Units.hs:278)
units-parser       > Warning: Text.Parse.Units: could not find link destinations for:
units-parser       >     Token UnitStringParser_UnitExp
units-parser       > Documentation created:
units-parser       > .stack-work/dist/x86_64-osx/Cabal-3.0.1.0/doc/html/units-parser/index.html,
units-parser       > .stack-work/dist/x86_64-osx/Cabal-3.0.1.0/doc/html/units-parser/units-parser.txt
units-parser       > copy/register
units-parser       > Installing library in /Users/pdejoux/.stack/snapshots/x86_64-osx/861454bd46483e99623797acad7e761a125bed3f708af95b7ac258877dc7020e/8.8.3/lib/x86_64-osx-ghc-8.8.3/units-parser-0.1.0.0-1r56BVImJ35E9Rfd3bjjcL
units-parser       > Registering library for units-parser-0.1.0.0..
uom-plugin         > configure (lib)
Configuring uom-plugin-0.3.0.0...
uom-plugin         > build (lib)
Preprocessing library for uom-plugin-0.3.0.0..
Building library for uom-plugin-0.3.0.0..
[ 1 of 18] Compiling Data.UnitsOfMeasure.Internal
[ 2 of 18] Compiling Data.UnitsOfMeasure.Singleton
[ 3 of 18] Compiling Data.UnitsOfMeasure.Show
[ 4 of 18] Compiling Data.UnitsOfMeasure.Read
[ 5 of 18] Compiling GhcApi
[ 6 of 18] Compiling GhcApi.Compare
[ 7 of 18] Compiling GhcApi.Shim
[ 8 of 18] Compiling Data.UnitsOfMeasure.Plugin.NormalForm
[ 9 of 18] Compiling Data.UnitsOfMeasure.Plugin.Convert
[10 of 18] Compiling Data.UnitsOfMeasure.Plugin.Unify
[11 of 18] Compiling GhcApi.Wrap
[12 of 18] Compiling Data.UnitsOfMeasure.Plugin
[13 of 18] Compiling Data.UnitsOfMeasure.Convert
[14 of 18] Compiling Data.UnitsOfMeasure.TH
[15 of 18] Compiling Data.UnitsOfMeasure
[16 of 18] Compiling Data.UnitsOfMeasure.Tutorial
[17 of 18] Compiling Data.UnitsOfMeasure.Defs
[18 of 18] Compiling Paths_uom_plugin
uom-plugin         > haddock
Preprocessing library for uom-plugin-0.3.0.0..
Running Haddock on library for uom-plugin-0.3.0.0..
Haddock coverage:
Warning: '-' is ambiguous. It is defined
    * in ‘GHC.Num’
    * in ‘GHC.TypeNats’
    You may be able to disambiguate the identifier by qualifying it or
    by specifying the type/value namespace explicitly.
    Defaulting to the one defined in ‘GHC.TypeNats’
Warning: '/:' is ambiguous. It is defined
    * at src/Data/UnitsOfMeasure/Internal.hs:182:14
    * at src/Data/UnitsOfMeasure/Internal.hs:88:1
    You may be able to disambiguate the identifier by qualifying it or
    by specifying the type/value namespace explicitly.
    Defaulting to the one defined at src/Data/UnitsOfMeasure/Internal.hs:88:1
Warning: '*:' is ambiguous. It is defined
    * at src/Data/UnitsOfMeasure/Internal.hs:157:14
    * at src/Data/UnitsOfMeasure/Internal.hs:82:1
    You may be able to disambiguate the identifier by qualifying it or
    by specifying the type/value namespace explicitly.
    Defaulting to the one defined at src/Data/UnitsOfMeasure/Internal.hs:82:1
 100% ( 37 / 37) in 'Data.UnitsOfMeasure.Internal'
 100% ( 10 / 10) in 'Data.UnitsOfMeasure.Singleton'
 100% (  3 /  3) in 'Data.UnitsOfMeasure.Show'
 100% (  6 /  6) in 'Data.UnitsOfMeasure.Read'
Warning: 'Unit' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'Base' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'One' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'Unpack' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'UnitSyntax' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: '(~~)' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: '(~~)' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
 100% (  2 /  2) in 'Data.UnitsOfMeasure.Plugin'
haddock: panic! (the 'impossible' happened)
  (GHC version 8.8.3 for x86_64-apple-darwin):
        mkPluginUsage: file not found

Data.UnitsOfMeasure.Plugin /var/folders/31/zb67fgcs66q4n3cs_wn134zr0000gn/T/.haddock-29062/Data/UnitsOfMeasure/Plugin.o
Call stack:
    CallStack (from HasCallStack):
      callStackDoc, called at compiler/utils/Outputable.hs:1159:37 in ghc:Outputable
      pprPanic, called at compiler/deSugar/DsUsage.hs:235:15 in ghc:DsUsage

Please report this as a GHC bug:  https://www.haskell.org/ghc/reportabug

Completed 5 action(s).

--  While building package uom-plugin-0.3.0.0 using:
      /Users/pdejoux/.stack/setup-exe-cache/x86_64-osx/Cabal-simple_mPHDZzAJ_3.0.1.0_ghc-8.8.3 --builddir=.stack-work/dist/x86_64-osx/Cabal-3.0.1.0 haddock --html --hoogle --html-location=../$pkg-$version/ --haddock-option=--hyperlinked-source --haddock-option=--quickjump
    Process exited with code: ExitFailure 1
```

**Comment by adamgundry** [comment]
This is a GHC/Haddock bug which is fixed in 9.4.x.
